// @zappar/zappar/lib/facemesh.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const zappar_1 = require("./zappar");
class FaceMesh {
    constructor() {
        this._z = zappar_1.z();
        this._impl = this._z.face_mesh_create();
    }
    destroy() {
        this._z.face_mesh_destroy(this._impl);
    }
    load(src, fillMouth, fillEyeLeft, fillEyeRight, fillNeck) {
        return __awaiter(this, void 0, void 0, function* () {
            if (typeof src === "string") {
                src = yield (yield fetch(src)).arrayBuffer();
            }
            this._z.face_mesh_load_from_memory(this._impl, src, fillMouth || false, fillEyeLeft || false, fillEyeRight || false, fillNeck || false);
        });
    }
    loadDefault() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._z.face_mesh_load_default(this._impl);
        });
    }
    loadDefaultFace(fillMouth, fillEyeLeft, fillEyeRight) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._z.face_mesh_load_default_face(this._impl, fillMouth || false, fillEyeLeft || false, fillEyeRight || false);
        });
    }
    loadDefaultFullHeadSimplified(fillMouth, fillEyeLeft, fillEyeRight, fillNeck) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._z.face_mesh_load_default_full_head_simplified(this._impl, fillMouth || false, fillEyeLeft || false, fillEyeRight || false, fillNeck || false);
        });
    }
    updateFromFaceAnchor(f, mirror) {
        this._z.face_mesh_update(this._impl, f.identity, f.expression, mirror || false);
    }
    updateFromIdentityExpression(identity, expression, mirror) {
        this._z.face_mesh_update(this._impl, identity, expression, mirror || false);
    }
    get vertices() {
        return this._z.face_mesh_vertices(this._impl);
    }
    get indices() {
        return this._z.face_mesh_indices(this._impl);
    }
    get uvs() {
        return this._z.face_mesh_uvs(this._impl);
    }
    get normals() {
        return this._z.face_mesh_normals(this._impl);
    }
    _getImpl() {
        return this._impl;
    }
}
exports.FaceMesh = FaceMesh;
