// @zappar/zappar-threejs/lib/defaultpipeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Zappar = require("@zappar/zappar");
const defaultPipeline = new Zappar.Pipeline();
function getDefaultPipeline() {
    return defaultPipeline;
}
exports.getDefaultPipeline = getDefaultPipeline;
class ImageTracker extends Zappar.ImageTracker {
    constructor(targetFile, pipeline) {
        super(pipeline || defaultPipeline, targetFile);
    }
}
exports.ImageTracker = ImageTracker;
class BarcodeFinder extends Zappar.BarcodeFinder {
    constructor(pipeline) {
        super(pipeline || defaultPipeline);
    }
}
exports.BarcodeFinder = BarcodeFinder;
class FaceTracker extends Zappar.FaceTracker {
    constructor(pipeline) {
        super(pipeline || defaultPipeline);
    }
}
exports.FaceTracker = FaceTracker;
class InstantWorldTracker extends Zappar.InstantWorldTracker {
    constructor(pipeline) {
        super(pipeline || defaultPipeline);
    }
}
exports.InstantWorldTracker = InstantWorldTracker;
class CameraSource extends Zappar.CameraSource {
    constructor(deviceId, pipeline) {
        super(pipeline || defaultPipeline, deviceId);
    }
}
exports.CameraSource = CameraSource;
class HTMLElementSource extends Zappar.HTMLElementSource {
    constructor(elm, pipeline) {
        super(pipeline || defaultPipeline, elm);
    }
}
exports.HTMLElementSource = HTMLElementSource;
function glContextSet(gl) {
    defaultPipeline.glContextSet(gl);
}
exports.glContextSet = glContextSet;
exports.onFrameUpdate = defaultPipeline.onFrameUpdate;
