// @zappar/zappar-threejs/lib/faceanchorgroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const THREE = require("three");
const camera_1 = require("./camera");
class FaceAnchorGroup extends THREE.Group {
    constructor(_camera, faceTracker, anchorId) {
        super();
        this._camera = _camera;
        this.faceTracker = faceTracker;
        this.anchorId = anchorId;
        this.isReady = false;
        // We'll be updating this Group's matrix ourselves from the Zappar library
        this.matrixAutoUpdate = false;
    }
    updateMatrixWorld(force) {
        if (!this.currentAnchor || !this.faceTracker.visible.has(this.currentAnchor)) {
            // No current anchor, or current anchor isn't visible
            if (this.anchorId) {
                this.currentAnchor = this.faceTracker.anchors.get(this.anchorId);
            }
            else {
                this.currentAnchor = this.faceTracker.visible.values().next().value;
            }
        }
        if (this.currentAnchor) {
            this.matrix.fromArray(this.currentAnchor.pose(this._camera.rawPose, this._camera.currentMirrorMode === camera_1.CameraMirrorMode.Poses));
        }
        super.updateMatrixWorld(force);
    }
}
exports.FaceAnchorGroup = FaceAnchorGroup;
